File: c/mt76_histogram_acquisition.c
/* SPDX-License-Identifier: BSD-3-Clause-Clear */
#include "mt76_histogram_acquisition.h"
#include <string.h>

static const uint32_t addresses[2][4]={{0x83082004,0x83088230,0x83088234,0},
                                      {0x83082004,0x83088230,0x83092004,0x83098230}};
static const uint32_t masks[2][4]={{7,1U<<29,0x30000,0},{7,1U<<29,7,1U<<29}};
static unsigned count(const mt_histogram_guard_t *g) { return g->chip==MT_CHIP_MT7925 ? 4 : 3; }
static int read_control(mt_histogram_guard_t *g, unsigned i, uint32_t *value) {
    return g->io.registers.read(g->io.registers.ctx,addresses[g->chip][i],value) || *value==UINT32_MAX ? -1 : 0;
}
static int set(mt_histogram_guard_t *g, unsigned i, uint32_t bits) {
    uint32_t word, check, mask=masks[g->chip][i];
    if (read_control(g,i,&word) || g->io.registers.write(g->io.registers.ctx,addresses[g->chip][i],(word&~mask)|bits) ||
        read_control(g,i,&check) || (check&mask)!=bits) return -1;
    return 0;
}
static uint32_t channel(const mt7921_dev_t *dev) {
    return (uint32_t)dev->tuned_band<<24|(uint32_t)dev->tuned_control<<16|
           (uint32_t)dev->tuned_center<<8|dev->tuned_width;
}
int mt_histogram_begin(mt_histogram_guard_t *g, int chip, mt_histogram_io_t io) {
    if (!g || g->active || (chip!=MT_CHIP_MT7921 && chip!=MT_CHIP_MT7925) || !io.registers.read || !io.registers.write ||
        (chip==MT_CHIP_MT7925 && !io.activate)) return -1;
    mt_histogram_guard_t next={.io=io,.chip=chip,.needs_reload=g->needs_reload};
    for (unsigned i=0;i<count(&next);i++) {
        uint32_t word;
        if (read_control(&next,i,&word)) return -1;
        next.saved[i]=word&masks[chip][i];
        if (next.saved[i] && !(chip==MT_CHIP_MT7921 && i==2)) return -1;
    }
    *g=next; g->active=g->pending=g->needs_reload=true;
    g->command_open_us=mt_radio_monotonic_us();
    if (chip==MT_CHIP_MT7925) {
        if (io.activate(io.command_ctx)) return -1;
    } else {
        if (set(g,0,0)) return -1;
        const uint32_t bits[3]={0,1U<<29,0};
        for (unsigned i=0;i<3;i++) {
            uint32_t word;
            if (read_control(g,1,&word) || io.registers.write(io.registers.ctx,addresses[0][1],(word&~(1U<<29))|bits[i])) return -1;
        }
        uint32_t word;
        if (read_control(g,1,&word) || (word&(1U<<29))) return -1;
        for (unsigned i=0;i<11;i++)
            if (io.registers.read(io.registers.ctx,0x83088600+4*i,&word) || word) return -1;
        if (set(g,2,0x30000) || set(g,0,5)) return -1;
    }
    g->command_closed_us=mt_radio_monotonic_us();
    g->ready=true;
    return 0;
}
static int activate(void *ctx) {
    mt7921_dev_t *dev=ctx;
    uint8_t request[8], reply[256]; uint32_t len=sizeof(reply),status;
    if (mt_histogram_request(dev->usb.chip,request,sizeof(request))<0 || mt7921_uni_option(dev->mcu.prof,0x36,false)!=7 ||
        mt7921_mcu_uni(&dev->mcu,0x36,request,sizeof(request),true,reply,&len,1000) ||
        mt_histogram_ack(dev->usb.chip,reply,len,dev->mcu.msg_seq,&status) || status) return -1;
    return 0;
}
int mt_histogram_begin_device(mt7921_dev_t *dev, mt_histogram_guard_t *g) {
    if (!dev || !g || g->active || !dev->tuned || dev->tuned_width!=20 || dev->tuned_center!=dev->tuned_control ||
        !((dev->tuned_band==0 && (dev->tuned_control==1 || dev->tuned_control==6 || dev->tuned_control==11)) ||
          (dev->tuned_band==1 && dev->tuned_control==36))) return -1;
    if (dev->usb.chip==MT_CHIP_MT7925 && (!dev->mcu.prof || mt7921_uni_option(dev->mcu.prof,0x36,false)!=7)) return -1;
    mt_histogram_io_t io={mt_radio_device_io(dev),dev,activate};
    int rc=mt_histogram_begin(g,dev->usb.chip,io);
    if (g->active) { g->dev=dev; g->channel_key=channel(dev); }
    return rc;
}
int mt_histogram_finish(mt_histogram_guard_t *g, const uint8_t *raw, size_t len, mt_histogram_sample_t *out) {
    if (!g || !g->active || !g->pending || !g->ready || !out || (g->dev && (!g->dev->tuned || channel(g->dev)!=g->channel_key))) return -1;
    bool modern=g->chip==MT_CHIP_MT7925;
    mt_histogram_sample_t result={0};
    if (modern ? mt_histogram_event(g->chip,raw,len,&result.bins) : (raw!=NULL || len!=0)) return -1;
    result.command_open_us=g->command_open_us; result.command_closed_us=g->command_closed_us;
    result.read_open_us=mt_radio_monotonic_us();
    if (modern) {
        for (unsigned i=0;i<4;i++) {
            uint32_t word;
            if (read_control(g,i,&word) || (word&masks[1][i])) return -1;
        }
        for (unsigned v=0;v<2;v++) for (unsigned i=0;i<11;i++) {
            uint32_t word;
            if (g->io.registers.read(g->io.registers.ctx,0x83001000+v*0x10000+4*i,&word) || word!=result.bins.bins[v][i]) return -1;
        }
    } else {
        if (set(g,0,0)) return -1;
        uint8_t words[44];
        for (unsigned i=0;i<11;i++) {
            uint32_t word;
            if (g->io.registers.read(g->io.registers.ctx,0x83088600+4*i,&word)) return -1;
            for (unsigned j=0;j<4;j++) words[4*i+j]=(uint8_t)(word>>(8*j));
        }
        if (mt_histogram_legacy(g->chip,words,sizeof(words),&result.bins)) return -1;
    }
    result.read_closed_us=mt_radio_monotonic_us();
    g->pending=g->ready=false; *out=result; return 0;
}
int mt_histogram_restore(mt_histogram_guard_t *g) {
    if (!g) return -1;
    if (!g->active) return 0;
    g->ready=false;
    if (g->chip==MT_CHIP_MT7925 && g->pending) return -1;
    int error=0;
    if (g->chip==MT_CHIP_MT7921 && set(g,0,0)) error=-1;
    const unsigned order[2][4]={{1,2,0,0},{1,3,0,2}};
    for (unsigned j=0;j<count(g);j++) {
        unsigned i=order[g->chip][j];
        if (set(g,i,g->saved[i])) error=-1;
    }
    if (!error) g->active=g->pending=false;
    return error;
}

File: mt76_histogram_acquisition.py
# SPDX-License-Identifier: BSD-3-Clause-Clear
"""Opt-in histogram control guard; execute methods through session.call.

One histogram owner per device; no retune, reset or out-of-band controls during
the guard lifetime. Caller routes events and bounds elapsed acquisition time.
Shared histories are destroyed. STOP/restoration cannot recover those histories.
"""

from __future__ import annotations

import struct
import time
from dataclasses import dataclass

from mt76_histogram import (
    HistogramBins,
    build_histogram_request,
    parse_histogram_ack,
    parse_histogram_event,
    parse_legacy_histogram,
)

_CONTROL = 0x83082004
_RESET = 0x83088230
_OPTIONS = 0x83088234
_OLD_MASKS = {_CONTROL: 7, _RESET: 1 << 29, _OPTIONS: 0x30000}
_NEW_MASKS = {_CONTROL: 7, _RESET: 1 << 29, 0x83092004: 7, 0x83098230: 1 << 29}
_CHANNELS = {("2.4GHz", channel, channel, 20) for channel in (1, 6, 11)} | {("5GHz", 36, 36, 20)}


@dataclass(frozen=True)
class HistogramSample:
    bins: HistogramBins
    command_open_us: int
    command_closed_us: int
    read_open_us: int
    read_closed_us: int
    coverage_fraction: float | None = None
    sample_period_ns: int | None = None


class HistogramGuard:
    """begin -> bounded caller acquisition -> finish -> restore.

    MT7925 finish requires its complete asynchronous event, then verifies stopped
    controls and matching timer banks. There is no known timer-cancel command:
    restore refuses while pending. Stop the USB worker and reload firmware if a
    pending modern acquisition fails/times out. Legacy restore can stop its engine.
    Restore failures retain active for retry. needs_reload stays true after any
    activation; explicitly reload after the experiment, never under a live worker.
    """

    def __init__(self, dev):
        if dev.CHIP not in ("mt7921", "mt7925"):
            raise ValueError("unsupported histogram chip")
        self.dev = dev
        self.modern = dev.CHIP == "mt7925"
        self._masks = _NEW_MASKS if self.modern else _OLD_MASKS
        self.active = self.pending = self.ready = self.needs_reload = False
        self._saved = {}
        self.command_open_us = self.command_closed_us = 0

    def _read_control(self, address):
        word = self.dev.rr(address)
        if type(word) is not int or not 0 <= word < 0xFFFFFFFF:
            raise RuntimeError("invalid histogram control read")
        return word

    def _set(self, address, bits):
        mask = self._masks[address]
        word = self._read_control(address)
        self.dev.wr(address, word & ~mask | bits)
        if self._read_control(address) & mask != bits:
            raise RuntimeError("histogram control readback mismatch")

    def begin(self):
        if self.active:
            raise RuntimeError("histogram guard already active")
        channel = getattr(self.dev, "_capture_channel", None)
        if channel not in _CHANNELS:
            raise ValueError("histogram requires a qualified20MHz channel")
        if self.modern and self.dev.uni_option(0x36) != 7:
            raise ValueError("histogram requires SET_ACK option7")
        saved = {
            address: self._read_control(address) & mask for address, mask in self._masks.items()
        }
        if any(bits for address, bits in saved.items() if address != _OPTIONS):
            raise RuntimeError("histogram already enabled or reset asserted")
        self._saved, self._channel = saved, channel
        self.active = self.pending = self.needs_reload = True
        self.ready = False
        self.command_open_us = time.monotonic_ns() // 1000
        if self.modern:
            raw = self.dev.mcu_uni(0x36, build_histogram_request("mt7925"), timeout=1000)
            if parse_histogram_ack("mt7925", raw, self.dev.msg_seq):
                raise RuntimeError("histogram command rejected")
        else:
            self._set(_CONTROL, 0)
            for bits in (0, 1 << 29, 0):
                word = self._read_control(_RESET)
                self.dev.wr(_RESET, word & ~(1 << 29) | bits)
            if self._read_control(_RESET) & (1 << 29):
                raise RuntimeError("histogram reset remained asserted")
            if any(self.dev.rr(0x83088600 + 4 * i) for i in range(11)):
                raise RuntimeError("histogram did not reset")
            self._set(_OPTIONS, 0x30000)
            self._set(_CONTROL, 5)
        self.command_closed_us = time.monotonic_ns() // 1000
        self.ready = True

    def finish(self, raw_event=None):
        if not self.active or not self.pending or not self.ready:
            raise RuntimeError("no pending histogram acquisition")
        if getattr(self.dev, "_capture_channel", None) != self._channel:
            raise RuntimeError("histogram channel changed")
        if not self.modern and raw_event is not None:
            raise ValueError("legacy histogram does not use an event")
        if self.modern and raw_event is None:
            raise ValueError("modern histogram requires its completed event")
        report = parse_histogram_event("mt7925", raw_event) if self.modern else None
        opened = time.monotonic_ns() // 1000
        if self.modern:
            if any(self._read_control(a) & mask for a, mask in self._masks.items()):
                raise RuntimeError("histogram timer has not stopped")
            observed = tuple(
                tuple(self.dev.rr(base + 4 * i) for i in range(11))
                for base in (0x83001000, 0x83011000)
            )
            if observed != report.bins:
                raise RuntimeError("histogram event/bank mismatch")
        else:
            self._set(_CONTROL, 0)
            words = tuple(self.dev.rr(0x83088600 + 4 * i) for i in range(11))
            report = parse_legacy_histogram("mt7921", struct.pack("<11I", *words))
        self.pending = False
        self.ready = False
        return HistogramSample(
            report,
            self.command_open_us,
            self.command_closed_us,
            opened,
            time.monotonic_ns() // 1000,
        )

    def restore(self):
        if not self.active:
            return
        self.ready = False
        if self.modern and self.pending:
            raise RuntimeError("pending histogram timer requires firmware reload")
        errors = []
        # Stop legacy first; restore enable bits last. Attempt every saved mask.
        if not self.modern:
            try:
                self._set(_CONTROL, 0)
            except Exception as exc:
                errors.append(exc)
        order = [a for a in self._masks if a not in (_CONTROL, 0x83092004)]
        order += [a for a in self._masks if a in (_CONTROL, 0x83092004)]
        for address in order:
            try:
                self._set(address, self._saved[address])
            except Exception as exc:
                errors.append(exc)
        if errors:
            raise RuntimeError("histogram restore failed") from errors[0]
        self.active = self.pending = False
